Lib/ufo2ft/kernFeatureWriter.py

```diff
@@ -99,32 +113,139 @@ def _collectClassKerning(self):
+        for (lClass, rClass), val in self.classPairKerning.items():
+            lGlyphs = leftClasses[lClass]
+            rGlyphs = rightClasses[rClass]
+            nlGlyphs, nrGlyphs = set(), set()
+            for lGlyph in lGlyphs:
+                for rGlyph in rGlyphs:
+                    pair = lGlyph, rGlyph
+                    if pair not in seen:
+                        nlGlyphs.add(lGlyph)
+                        nrGlyphs.add(rGlyph)
+                        seen[pair] = val
+            nlClass, nrClass = lClass, rClass
+            if nlGlyphs != set(lGlyphs):
+                nlClass = self._liststr(sorted(nlGlyphs))
+            if nrGlyphs != set(rGlyphs):
+                nrClass = self._liststr(sorted(nrGlyphs))
+            if nlClass != lClass or nrClass != rClass:
+                self.classPairKerning[nlClass, nrClass] = val
+                del self.classPairKerning[lClass, rClass]
 
-        self._collectClassKerning()
+    def _addGlyphClasses(self, lines):
+        """Add glyph classes for the input font's groups."""
 
-        # write the glyph classes
-        lines = []
-        self._addGlyphClasses(lines)
-        lines.append("")
+        for key, members in sorted(self.leftUfoClasses.items() +
+                                   self.rightUfoClasses.items()):
+            lines.append("%s = [%s];" % (key, " ".join(members)))
 
-        # write the feature
-        lines.append("feature kern {")
-        self._addKerning(lines)
-        lines.append("    subtable;")
-        self._addKerning(lines, self.leftClassKerning, enum=True)
-        lines.append("    subtable;")
-        self._addKerning(lines, self.rightClassKerning, enum=True)
-        lines.append("    subtable;")
-        self._addKerning(lines, self.classPairKerning)
-        lines.append("} kern;")
+    def _addKerning(self, lines, kerning, enum=False):
+        """Add kerning rules for a mapping of pairs to values."""
+
+        enum = "enum " if enum else ""
+        for (left, right), val in sorted(kerning.items()):
+            lines.append("    %spos %s %s %d;" % (enum, left, right, val))
+
+    def _liststr(self, glyphs):
+        """Return string representation of a list of glyph names."""
+
+        return "[%s]" % " ".join(glyphs)
+
+    def _isClassName(self, nameRe, name):
+        """Return whether a given name matches a given class name regex."""
+
+        return re.match(nameRe, name) is not None
 
-        # return the feature, unless it's empty
-        return "" if len([ln for ln in lines if ln]) == 5 else linesep.join(lines)
+    def _makeFeaClassName(self, name):
+        """Make a glyph class name which is legal to use in OTF syntax.
+
+        Ensures the name starts with "@" and only includes characters in
+        "A-Za-z0-9._", and isn't already defined.
+        """
+
+        if not name.startswith("@"):
+            name = "@" + name
+        name = re.sub(r"[^A-Za-z0-9._]", r"", name)
```

Comment:
Aren't you dropping the prefix @ when doing this? Maybe perform that instruction first then unconditionally add an @.